### PR TITLE
Driver/ziuhfli

### PR DESCRIPTION
**Pull request by WilliamHPNielsen**
feature: Add a driver for the ZI UHF-LI

Add a driver implementing basic control of demodulator settings and access to the sweeper.

Minor tasks still needing to be completed:
- [x] Coding style, typos, whitespace
- [x] Add example notebook 
- [x] Add more "front panel" parameters

Major tasks:
- [x] Add the scope.

@giulioungaretti @jenshnielsen

### CODE REVIEW

**Comment by AdriaanRol** [comment]
@WilliamHPNielsen . We are testing a new piece of hardware from ZI that is based on the UHF-LI. We are using alpha version firmware so I expect the driver to change a lot. Nonetheless, Niels Bultink has implemented a way to auto generate the qcodes driver based on the ZI API. 

I think we probably have some useful snippets for you in our repo: https://github.com/DiCarloLab-Delft/PycQED_py3/tree/2Q_gates_branch/pycqed/instrument_drivers/physical_instruments/ZurichInstruments . 

I'd be curious to know what you think of it.

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 1]
```diff
@@ -45,6 +45,8 @@ def __init__(self, name, instrument, **kwargs):
         super().__init__(name, names=('',), shapes=((1,),), **kwargs)
         self._instrument = instrument
 
+        print('We rock out at git!')
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: lol @WilliamHPNielsen  signature ? 🦄

Comment (WilliamHPNielsen) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: @giulioungaretti Boy better know! But actually, Jana promised to practice her git-fu by removing it before anyone saw it...

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 2]
```diff
@@ -0,0 +1,1925 @@
...
+    """
+
+    def __init__(self, name, device_ID, api_level=5, **kwargs):
+        """
+        Create an instance of the instrument.
+
+        Args:
+            name (str): The internal QCoDeS name of the instrument
+            device_ID (str): The device name as listed in the web server.
+            api_level (int): Compatibility mode of the API interface. Must be 5
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: mmhm ? if it must be five why it's an argument ?

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 3]
```diff
@@ -0,0 +1,1925 @@
...
+    QCoDeS driver for ZI UHF-LI.
+
+    Currently implementing demodulator settings and the sweeper functionality.
+
+    Requires ZI Lab One software to be installed on the computer running QCoDeS.
+    Furthermore, the Data Server and Web Server must be running and a connection
+    between the two must be made.
+
+    TODOs:
+        * Add the scope
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: Is it not added now ?

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 4]
```diff
@@ -0,0 +1,1925 @@
...
+        Args:
+            name (str): The internal QCoDeS name of the instrument
+            device_ID (str): The device name as listed in the web server.
+            api_level (int): Compatibility mode of the API interface. Must be 5
+              for the UHF.
+        """
+
+        super().__init__(name, **kwargs)
+        zisession = zhinst.utils.create_api_session(device_ID, api_level)
+        (self.daq, self.device, self.props) = zisession
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: why  a tuple ?

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 5]
```diff
@@ -0,0 +1,1925 @@
...
+
+        self.daq.setDebugLevel(3)
+        # create (instantiate) an instance of each module we will use
+        self.sweeper = self.daq.sweep()
+        self.sweeper.set('sweep/device', self.device)
+        #
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: ⚡️ that  hash

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 6]
```diff
@@ -0,0 +1,1925 @@
...
+                      'Output 1 Amplitude 4': 'V',
+                      'Output 1 Offset': 'V',
+                      'Output 2 Amplitude 8': 'V',
+                      'Output 2 Offset': 'V'
+                      }
+
+        self.add_parameter('sweeper_units',
+                           label='Units of sweep x-axis',
+                           get_cmd=self.sweeper_param.get,
+                           get_parser=lambda x:sweepunits[x])
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: does this really really work as expected ? @WilliamHPNielsen  thinking about late binding of lambdas

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 7]
```diff
@@ -0,0 +1,1925 @@
...
+                                     time is taken as wait time until the next
+                                     sweep point is recorded.
+                                     """
+                           )
+
+        self.add_parameter('sweeper_settlingtc',
+                           label='Sweep filter settling time',
+                           get_cmd=partial(self._sweep_getter,
+                                           'sweep/settling/tc'),
+                           unit='dim. less.',
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: this or empty ? I  may say that dimensionless has empty unit ? what you think ?

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 8]
```diff
@@ -0,0 +1,1925 @@
...
+                            label="Scope's mode: time or frequency domain.",
+                            set_cmd=partial(self._scope_setter, 1, 0,
+                                            'mode'),
+                            get_cmd=partial(self._scope_getter, 'mode'),
+                            val_mapping={'Time Domain': 1,
+                                         'Freq Domain FFT': 3},
+                            vals=vals.Enum('Time Domain', 'Freq Domain FFT')
+                            )
+
+        # 1: Channel 1 on, Channel 2 off. 2: Channel 1 off, Channel 2 on,
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: maybe make it 3 lines i.e
1
2
3
?

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 9]
```diff
@@ -0,0 +1,1925 @@
...
+                       'Aux In 1 Ch 2': 9,
+                       'Osc phi Demod 4': 10,
+                       'Osc phi Demod 8': 11,
+                       'AU Cartesian 1': 112,
+                       'AU Cartesian 2': 113,
+                       'AU Polar 1': 128,
+                       'AU Polar 2': 129,
+                       }
+
+        for demod in range(1,9):
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: definitely add a comment on the magic numbers  ❤️  ! 🦄

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py (hunk 3 above)
Comment (jana-d) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: yes

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py (hunk 4 above)
Comment (jana-d) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: why not?

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py (hunk 5 above)
Comment (jana-d) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: oups

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py (hunk 7 above)
Comment (jana-d) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: empty

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py (hunk 4 above)
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: self.daq, self.device, self.props = zisession

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py (hunk 6 above)
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: lovely

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py (hunk 7 above)
Comment (giulioungaretti) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: <3

**Pull request merged**